File: src/com/example/javas/medium/M008_AnagramChecker.java
/*
MEDIUM PROBLEM (Simplified for 25 min):
Check if two strings are anagrams

REQUIREMENTS:
- Return true if two strings are anagrams (contain same characters with same frequency)
- Case-insensitive
- Ignore spaces

APPROACH:
- Clean both strings (remove spaces, convert to lowercase)
- Use HashMap to count character frequencies
- Compare the two frequency maps

DATA STRUCTURES:
- HashMap<Character, Integer>: character -> frequency count

TIME COMPLEXITY: O(n + m) where n, m are string lengths
SPACE COMPLEXITY: O(n + m)
*/

package com.example.javas.medium;

import java.util.*;

public class M008_AnagramChecker {
    
    public static boolean isAnagram(String s1, String s2) {
        // TODO: Implement anagram checker
        // 1. Clean both strings: remove spaces, convert to lowercase
        // 2. If lengths differ, return false
        // 3. Create frequency map for first string
        // 4. Create frequency map for second string
        // 5. Compare the two maps using .equals()
        
        return false;
    }
    
    // Helper: create frequency map for a string
    private static HashMap<Character, Integer> getFrequencyMap(String s) {
        // TODO: Implement frequency counter
        // 1. Create HashMap
        // 2. Loop through string
        // 3. Use getOrDefault to increment count
        // 4. Return map
        
        return new HashMap<>();
    }
    
    // TEST CASES
    public static void main(String[] args) {
        System.out.println("\"listen\" & \"silent\" -> " + isAnagram("listen", "silent"));
        System.out.println("\"hello\" & \"world\" -> " + isAnagram("hello", "world"));
        System.out.println("\"The Eyes\" & \"They See\" -> " + isAnagram("The Eyes", "They See"));
        System.out.println("\"abc\" & \"def\" -> " + isAnagram("abc", "def"));
        System.out.println("\"a\" & \"a\" -> " + isAnagram("a", "a"));
    }
}

File: pythonX/medium/m008_anagram_checker.py
"""
M008: Anagram Checker

Summary:
Check if two strings are anagrams of each other.

Requirements:
Return True if strings are anagrams (same chars, same frequency).
Case-insensitive, ignore spaces.

Approach:
Clean strings, count character frequencies, compare.

Data Structures:
- dict[str, int]: character -> frequency

Big O Analysis:
Time Complexity: O(n + m) where n, m are string lengths
Space Complexity: O(n + m)
"""


def is_anagram(s1: str, s2: str) -> bool:
    """
    TODO: Check if strings are anagrams
    1. Clean both strings: remove spaces, convert to lowercase
    2. If lengths differ, return False
    3. Create frequency map for both strings
    4. Compare the two maps
    """
    pass


def get_frequency_map(s: str) -> dict[str, int]:
    """
    TODO: Create frequency map for string
    1. Create dict
    2. Loop through string
    3. Count each character
    4. Return dict
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Basic anagrams")
    print(f"'listen' & 'silent' -> {is_anagram('listen', 'silent')}")  # Expected: True
    print(f"'hello' & 'world' -> {is_anagram('hello', 'world')}")      # Expected: False
    
    print("\nTest 2: With spaces")
    print(f"'The Eyes' & 'They See' -> {is_anagram('The Eyes', 'They See')}")  # Expected: True
    
    print("\nTest 3: Not anagrams")
    print(f"'abc' & 'def' -> {is_anagram('abc', 'def')}")              # Expected: False
    
    print("\nTest 4: Same character")
    print(f"'a' & 'a' -> {is_anagram('a', 'a')}")                      # Expected: True
